src/components/MutateBookmark/MutateBookmark.tsx

```diff
@@ -0,0 +1,201 @@
+import { useEffect, useState } from "react";
+import { useForm } from "react-hook-form";
+import { v4 as uuidV4 } from "uuid";
+import { useSettingsUtilities } from "../../hooks/useSettingsUtilities";
+import { IBookmark } from "../../types/bookmarks";
+import { DropdownSelect } from "../DropdownSelect/DropdownSelect";
+import { InputWithTitle } from "../InputWithTitle/InputWithTitle";
+type Props = {
+  type: "Add" | "Edit";
+  objectId?: string;
+  folderName?: string;
+  objectName: string;
+  setPageSettings: (
+    obj: {
+      type: "Add" | "Edit";
+      objectName: "Bookmark" | "Folder";
+      objectId?: string;
+    } | null
+  ) => void;
+};
+
+type InputType = {
+  Name: string;
+  URL: string;
+  Description: string;
+  ParentFolder: string | null;
+  Id: string;
+};
+
+export const MutateBookmark = ({
+  type,
+  objectId,
+  objectName,
+  setPageSettings,
+}: Props) => {
+  const bookmarkUtilities = useSettingsUtilities();
+  const [gotParentFolder, setGotParentFolder] = useState<boolean>(false);
+  const [parentFolders, setParentFolders] = useState<
+    {
+      value: string | null;
+      label: string;
+    }[]
+  >([]);
+
+  useEffect(() => {
+    if (!bookmarkUtilities || gotParentFolder) return;
+
+    const parentFoldersFromFunc = bookmarkUtilities.getParentFolders();
+
+    setParentFolders([
+      ...parentFoldersFromFunc.map((e) => ({
+        value: e.Id,
+        label: e.Name === "Root" ? "No Parent" : e.Name,
+      })),
+    ]);
+
+    setGotParentFolder(true);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [bookmarkUtilities]);
+
+  const {
+    register,
+    formState: { errors },
+    handleSubmit,
+    reset,
+    watch,
+    setValue,
+    setError,
+  } = useForm<InputType>();
+
+  useEffect(() => {
+    setError("URL", {
+      type: "manual",
+      message: "Valid URL is required",
+    });
+  }, [setError]);
+
+  useEffect(() => {
+    if (objectName !== "Folder") return;
+
+    const folders = bookmarkUtilities?.getParentFolders();
+
+    setValue(
+      "ParentFolder",
+      folders?.find((e) => e.Name === "Root")?.Id as string
+    );
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [objectName]);
+
+  useEffect(() => {
+    if (type !== "Edit" || !objectId || !bookmarkUtilities) return;
+
+    const bookmark = bookmarkUtilities
+      .getBookmarks()
+      .find((e) => e.Id === objectId);
+
+    reset(bookmark);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [type, objectId]);
+
+  useEffect(() => {
+    if (type !== "Add" || !bookmarkUtilities) return;
+
+    const rootFolderId = bookmarkUtilities
+      .getParentFolders()
+      .find((e) => e.Name === "Root")?.Id;
+
+    reset({
+      ParentFolder: rootFolderId,
+      Id: uuidV4(),
+    });
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [type]);
+
+  if (!bookmarkUtilities) return <></>;
+
+  const submit = (data: IBookmark) => {
```

Comment:
use `useCallback` to memoize the function

Reply:
It's a submit function, this is not needed.

Reply:
Now this function is recalculated at each render